lib/Basic/Unix/TaskQueue.inc

```diff
@@ -249,45 +245,48 @@ bool Task::execute() {
   return false;
 }
 
-static bool readFromAPipe(int Pipe, std::string &Output,
-                          UnifiedStatsReporter *Stats,
-                          bool UntilEnd) {
+/// Read the data in \p Pipe, and append it to \p Output.
+/// \p Pipe must be in blocking mode, and must contain unread data.
+/// If \p UntilEnd is true, keep reading, and possibly hanging, till the pipe is
+/// closed. If \p UntilEnd is false, just read once. Return true if error
+static bool readFromAPipe(std::string &Output, int Pipe,
+                          UnifiedStatsReporter *Stats, bool UntilEnd) {
   char outputBuffer[1024];
   ssize_t readBytes = 0;
   while ((readBytes = read(Pipe, outputBuffer, sizeof(outputBuffer))) != 0) {
     if (readBytes < 0) {
       if (errno == EINTR)
         // read() was interrupted, so try again.
+        // FIXME: There should probably be a counter to break out of this loop
+        // if there are more than some number of EINTRs.
         continue;
       return true;
     }
-
     Output.append(outputBuffer, readBytes);
     if (Stats)
       Stats->getDriverCounters().NumDriverPipeReads++;
     if (!UntilEnd)
       break;
   }
-
   return false;
 }
 
 bool Task::readFromPipes(bool UntilEnd) {
-  bool Ret = readFromAPipe(Pipe, Output, Stats, UntilEnd);
+  bool Ret = readFromAPipe(Output, Pipe, Stats, UntilEnd);
```

Comment:
I wanted to separate the in-args from the out-args, but then Stats came along.
What order do you like? Why? I can see `Pipe, UntilEnd, Stats, Output` (chronological) or `Output, Stats, Pipe, UntilEnd` (output, then input). I don't feel super-strongly though I tend to try to avoid out-args in general.

Reply:
Ok. Was just wondering, makes sense.

Reply:
Thanks.